File: example_clients/javascript/tokenbucket_client.js
const net = require('net');

class TokenBucketClient {
    constructor(hostname, port) {
        this.server = net.createConnection({ host: hostname, port: port });
        this.server.setEncoding('utf8');
    }

    async consume(bucketName, callback) {
        this.server.write(`CONSUME ${bucketName}\n`);
        for await (const data of this.server) {
            const [status, ...rest] = data.split(' ');

            switch (status) {
                case 'OK':
                    this.server.end();
                    return callback();
                case 'WAIT':
                    const [sleepTime, ..._] = rest;
                    await new Promise(resolve => setTimeout(resolve, sleepTime * 1000));
                    this.server.write(`CONSUME ${bucketName}\n`);
                    break;
                default:
                    this.server.end();
                    throw new Error(`Unknown response from server: ${data}`);
            }
        }
    }
}

// Example usage:
/*
const client = new TokenBucketClient('localhost', 2000);
client.consume('foo', () => {
  // do something
});
*/

File: example_clients/java/TokenBucketClient.java
import java.io.*;
import java.net.Socket;
import java.util.concurrent.TimeUnit;

public class TokenBucketClient {
    private String hostname;
    private int port;

    public TokenBucketClient(String hostname, int port) {
        this.hostname = hostname;
        this.port = port;
    }

    public void consume(String bucketName, Runnable callback) throws IOException, InterruptedException {
        try (Socket socket = new Socket(hostname, port);
             PrintWriter out = new PrintWriter(socket.getOutputStream(), true);
             BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()))) {

            while (true) {
                out.println("CONSUME " + bucketName);
                String response = in.readLine();
                String[] parts = response.split(" ");

                switch (parts[0]) {
                    case "OK":
                        callback.run();
                        return;
                    case "WAIT":
                        long sleepTime = (long) (Double.parseDouble(parts[1]) * 1000);
                        Thread.sleep(sleepTime);
                        break;
                    default:
                        throw new RuntimeException("Unknown response from server: " + response);
                }
            }
        }
    }

    // Example usage:
    // TokenBucketClient client = new TokenBucketClient("localhost", 2000);
    // client.consume("foo", () -> { /* do something */ });
}
